yt_dlp/YoutubeDL.py

```diff
@@ -1423,9 +1425,42 @@ def create_key(outer_mobj):
 
         return EXTERNAL_FORMAT_RE.sub(create_key, outtmpl), TMPL_DICT
 
-    def evaluate_outtmpl(self, outtmpl, info_dict, *args, **kwargs):
+    def evaluate_outtmpl(self, outtmpl, info_dict, *args, trim_filename=False, **kwargs):
         outtmpl, info_dict = self.prepare_outtmpl(outtmpl, info_dict, *args, **kwargs)
-        return self.escape_outtmpl(outtmpl) % info_dict
+        if not trim_filename:
+            return self.escape_outtmpl(outtmpl) % info_dict
+
+        ext_suffix = '.%(ext\0s)s'
+        suffix = ''
+        if outtmpl.endswith(ext_suffix):
+            outtmpl = outtmpl[:-len(ext_suffix)]
+            suffix = ext_suffix % info_dict
+        outtmpl = self.escape_outtmpl(outtmpl)
+        filename = outtmpl % info_dict
+
+        def parse_trim_file_name(trim_file_name):
+            if trim_file_name is None or trim_file_name == 'notrim':
+                return 0, None
+            mobj = re.match(r'(?:(?P<length>\d+)(?P<mode>b|c)?|notrim)', trim_file_name)
+            return int(mobj.group('length')), mobj.group('mode') or 'c'
+
+        max_file_name, mode = parse_trim_file_name(self.params.get('trim_file_name'))
+        if max_file_name == 0:
+            # no maximum
+            return filename + suffix
+
+        encoding = self.params.get('filesystem_encoding') or sys.getfilesystemencoding()
+
+        def trim_filename(name: str):
+            if mode == 'b':
+                name = name.encode(encoding)
+                name = name[:max_file_name]
+                return name.decode(encoding, 'ignore')
+            else:
+                return name[:max_file_name]
+
+        filename = os.path.join(*map(trim_filename, Path(filename).parts))
```

Comment:
Doing this will mean that directory names are unnecessarily truncated to say 240 chars instead of 255 if we set a --trim-filenames small enough to avoid errors with the longest possible extension. It _might_ make sense to have --trim-filenames take the extension into account (so we can set --trim-filenames=255c) and then subtract the maximum length extension only for the last path component.